volatility3/framework/symbols/linux/utilities/modules.py

```diff
@@ -210,77 +225,173 @@ def get_kernel_module_info(
         context: interfaces.context.ContextInterface,
-        kernel_name: str,
-        run_hidden_modules: bool = True,
+        kernel_module_name: str,
+        caller_wanted_sources: List[str],
         flatten: bool = True,
     ) -> Dict[str, List[ModuleInfo]]:
         """Run module scanning plugins and aggregate the results. It is designed
         to not operate any inter-plugin results triage.
 
+        Rules for `caller_wanted_sources`:
+
+            If `Modules.all_sources_identifier` is specified then every source will be populated
+
+            If `Modules.source_hidden_identifier` is in the list, then at least one other sources must be
+            specified so a comparison will be populated
+
+            If empty or an invalid source is specified then a ValueError is thrown
+
         Args:
-            run_hidden_modules: specify if the hidden_modules plugin should be run
+            called_wanted_sources: The list of sources to gather modules.
+            flatten: Whether to de-duplicate modules across sources
         Returns:
             Dictionary mapping each plugin to its corresponding result
         """
 
-        kernel = context.modules[kernel_name]
+        module_gatherers = {
+            Modules.source_kernel_identifier: cls.get_kernel_module_info,
+            Modules.source_lsmod_identifier: cls._get_list_modules,
+            Modules.source_sysfs_identifier: cls._get_sysfs_modules,
+        }
+
+        kernel = context.modules[kernel_module_name]
 
         address_mask = context.layers[kernel.layer_name].address_mask
 
+        wanted_sources = Modules._validated_sources(caller_wanted_sources)
+
         run_results = {}
 
-        # the kernel module boundaries
-        run_results["kernel"] = [cls.get_kernel_module_info(context, kernel_name)]
+        run_hidden_modules = False
 
-        # lsmod
-        run_results["lsmod"] = []
+        # Special case hidden modules since it gathers modules on its own
+        if Modules.source_hidden_identifier in wanted_sources:
+            run_hidden_modules = True
+            wanted_sources.remove(Modules.source_hidden_identifier)
 
-        for module in cls.list_modules(context, kernel_name):
-            modinfo = cls.get_module_info_for_module(address_mask, module)
-            if modinfo:
-                run_results["lsmod"].append(modinfo)
+        # Walk each source, gathering modules
+        for wanted_source in wanted_sources:
+            run_results[wanted_source] = []
 
-        # check_modules
-        run_results["check_modules"] = []
+            gatherer = module_gatherers[wanted_source]
 
-        sysfs_modules: dict = cls.get_kset_modules(context, kernel_name)
+            # process each module coming from back the current source
+            for module in gatherer(context, kernel_module_name):
+                # the kernel sends back a ModuleInfo directly
+                if wanted_source == Modules.source_kernel_identifier:
```

Comment:
I realize this will be for efficiency, but it also prevents making the sources into an interface that others can add to.